File: OBlock/src/Module.cpp
#include "Module.h"

Module::Module(string code, string course, int maxCapacity) {
  this->code = code;
  this->course = course;
  this->maxCapacity = maxCapacity;
  this->enrolledStudents = vector<Student*>();
  this->requisiteModules = vector<Module*>();
  this->compatibleCourses = vector<string>();
}

string Module::getModuleCode() {
  return this->code;
}

string Module::getCourse() {
    return course;
}

int Module::getMaxCapacity() {
  return this->maxCapacity;
}

void Module::getMaxCapacity(int maxCapacity) {
  this->maxCapacity = maxCapacity;
}

vector<Student *> Module::getEnrolledStudents() {
  return this->enrolledStudents;
}

vector<Module *> Module::getRequisiteModules() {
  return this->requisiteModules;
}

vector<string> Module::getCompatibleCourses() {
  return this->compatibleCourses;
}

int Module::getEnrolledStudentIndex(Student* student) {
  for (int i = 0; i < this->enrolledStudents.size(); i++) {
    if (this->enrolledStudents[i] == student) {
      return i;
    }
  }
  return -1;
}

bool Module::addEnrolledStudent(Student* student) {
  if (this->enrolledStudents.size() < this->maxCapacity) {
    this->enrolledStudents.push_back(student);
    return true;
  }
  return false;
}

bool Module::removeEnrolledStudent(Student* student) {
  const int index = this->getEnrolledStudentIndex(student);
  if (index == -1) {
    return false;
  }
  this->enrolledStudents.erase(this->enrolledStudents.begin() + index);
  return true;
}

bool Module::isStudentEnrolled(Student* student) {
  if (this->enrolledStudents.size() > 0) {
    for (int i = 0; i < this->enrolledStudents.size(); i++) {
      if (this->enrolledStudents[i] == student) {
        return true;
      }
    }
  }
  return false;
}

void Module::addRequisiteModule(Module* module) {
  this->requisiteModules.push_back(module);
}

bool Module::removeRequisiteModule(Module* module) {
  if (this->requisiteModules.size() > 0) {
    for (int i = 0; i < this->requisiteModules.size(); i++) {
      if (this->requisiteModules[i] == module) {
        this->requisiteModules.erase(this->requisiteModules.begin() + i);
        return true;
      }
    }
  }
  return false;
}

void Module::removeOverbookings() {
  this->enrolledStudents.erase(this->enrolledStudents.begin() + (this->maxCapacity - 1));
}

File: OBlock/src/module.js
let moduleKeys = ['CS4123'];

class Module {
    constructor(code, maxCapacity, students, compatibleCourses, requisiteModules) {
        this.code = code;
        this.courses = compatibleCourses;
        this.maxCapacity = maxCapacity;
        this.students = students;
        this.requisiteModules = requisiteModules;
    }
}

async function createModule(event) {
    event.preventDefault();

    const code = document.getElementById('moduleCode').value;
    const capacity = document.getElementById('moduleCapacity').value;
    const compatibleCourses = document.getElementById('compatibleCourses').value.split(',');
    const requisites = document.getElementById('requisites').value.split(',');

    const accounts = await web3.eth.getAccounts();
    const address = accounts[0];

    await contract.methods.createModule(code, capacity).send({ from: address });

    for (let i = 0; i < compatibleCourses.length; i++) {
        await contract.methods.addCompatibleCourse(code, compatibleCourses[i]).send({ from: address });
    }

    for (let i = 0; i < requisites.length; i++) {
        await contract.methods.addRequisiteModule(code, requisites[i]).send({ from: address });
    }

    moduleKeys.push(code);
    displayModules();
}

async function displayModules() {
    const moduleTableBody = document.getElementById('moduleTableBody');
    moduleTableBody.innerHTML = '';

    for (let i = 0; i < moduleKeys.length; i++) {
        const module = await contract.methods.getModule(moduleKeys[i]).call();
        console.log(module);
        let row = document.createElement('tr');
        row.innerHTML = `
           <td>${module.code}</td>
           <td>${module.maxCapacity}</td>
           <td>${module.students.length}</td>
           <td class="scrollableList">
               <div>${module.students.map(student => `
                   <div style="border-bottom: 1px solid #aaa;">
                       <span>${student}</span>
                   </div>
               `).join('')}</div>
           </td>
           <td class="scrollableList">
               ${module.compatibleCourses.map(course => `<span>${course}</span>`).join('<br>')}
           </td>
           <td class="scrollableList">
               ${module.requisiteModules.map(course => `<span>${course}</span>`).join('<br>')}
           </td>
       `;
        moduleTableBody.appendChild(row);
    }
}